src/storage/store/free_chunk_map.cpp

```diff
@@ -0,0 +1,177 @@
+    FreeChunkLevel curLevel = GetChunkLevel(numPages);
+    KU_ASSERT(curLevel < MAX_FREE_CHUNK_LEVEL);
+
+    /* 2. return nullptr if have no entry for the given level and above */
+    if (maxAvailLevel < curLevel) {
+        return nullptr;
+    }
+
+    /* 3. Now, search an usable entry for given numPages and timestamp unitl exceeding the max level */
+    while (curLevel <= maxAvailLevel) {
+        /* if the level we are searching has no space to reuse, continue search the next level */
+        if (freeChunkList[curLevel] == nullptr) {
+            curLevel = static_cast<FreeChunkLevel>((curLevel + 1 < maxAvailLevel)? curLevel + 1:maxAvailLevel);
+            continue;
+        }
+
+        /* Search the current level for an entry with valid reuse timestamp */
+        KU_ASSERT(freeChunkList[curLevel] != nullptr);
+        FreeChunkEntry *curEntry = freeChunkList[curLevel];
+        FreeChunkEntry *lastSearchEntry = nullptr;
+        while (curEntry != nullptr) {
+            if (curEntry->numPages >= numPages) {
+                /* found a valid entry to reuse. Remove it from current linked list */
+                if (lastSearchEntry == nullptr) {
+                    /* the valid entry is the first entry in the L.L. */
+                    freeChunkList[curLevel] = curEntry->nextEntry;
+
+                    /* update maxAvailLevel if needed */
+                    /* ERICTODO: Change this to be a helper function */
+                    if (curLevel == maxAvailLevel && freeChunkList[curLevel] == nullptr) {
+                        FreeChunkLevel nextAvailLevel = INVALID_FREE_CHUNK_LEVEL;
+                        for (int i = curLevel; i >= FREE_CHUNK_LEVEL_0; i--) {
+                            if (freeChunkList[i] != nullptr) {
+                                nextAvailLevel = static_cast<FreeChunkLevel>(i);
+                                break;
+                            }
+                        }
+                        maxAvailLevel = nextAvailLevel;
+                    }
+                } else {
+                    /* need to unlink it from its parent */
+                    lastSearchEntry->nextEntry = curEntry->nextEntry;
+                }
+
+                curEntry->nextEntry = nullptr;
+                return curEntry;
+            }
+
+            /* Move to the next entry */
+            lastSearchEntry = curEntry;
+            curEntry = curEntry->nextEntry;
+        }
+    }
+    
+    /* No reusable chunk. Just return nullptr here */
+    return nullptr;
+}
+
+void FreeChunkMap::AddFreeChunk(common::page_idx_t pageIdx, common::page_idx_t numPages)
+{
+    KU_ASSERT(pageIdx != INVALID_PAGE_IDX && numPages != 0);
+
+    /* 1. Get the corresponding ChunkLevel numPages belongs to */
+    FreeChunkLevel curLevel = GetChunkLevel(numPages);
+    KU_ASSERT(curLevel < MAX_FREE_CHUNK_LEVEL);
+
+    /* 2. Create a new FreeChunkEntry */
+    FreeChunkEntry *entry = (FreeChunkEntry *)malloc(sizeof(FreeChunkEntry));
```

Comment:
Curious - why are we using malloc and free in C++ environment? Is there something about 'new' and 'delete' that doesn't work for our use case?